### PR TITLE
taint toleration filtering

### PR DESCRIPTION
**Pull request by AnthonyXu109**
Filter out clusters with untolerated taints.

### CODE REVIEW

File: pkg/plugins/tainttoleration/taint_toleration.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	// If the placement has no toleration, all clusters with taint should be filtered out.
+	if len(placement.Spec.Tolerations) == 0 {
+		clusterWithoutTaints := []*clusterapiv1.ManagedCluster{}
+		for _, cluster := range clusters {
+			if len(cluster.Spec.Taints) == 0 {
+				clusterWithoutTaints = append(clusterWithoutTaints, cluster)
+			}
+		}
+		return clusterWithoutTaints, nil
+	}
```
Comment (elgnay) [inline pkg/plugins/tainttoleration/taint_toleration.go:57]: This block is unnecessary.

File: pkg/plugins/tainttoleration/taint_toleration.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+		if len(toleration.Key) == 0 && toleration.Operator != clusterapiv1alpha1.TolerationOpExists {
+			return nil, errors.New("If the key is empty, operator must be Exists.\n")
+		}
+		if toleration.Operator == clusterapiv1alpha1.TolerationOpExists && len(toleration.Value) > 0 {
+			return nil, errors.New("If the operator is Exists, the value should be empty.\n")
+		}
+		if toleration.TolerationSeconds != nil && toleration.Effect != clusterapiv1.TaintEffectNoSelect && toleration.Effect != clusterapiv1.TaintEffectPreferNoSelect {
+			fmt.Println("Warning: TolerationSeconds would be ignored if Effect is not NoSelect/PreferNoSelect.")
+		}
+	}
```
Comment (elgnay) [inline pkg/plugins/tainttoleration/taint_toleration.go:47]: @haoqing0110 Is it possible for a filter to add a `PlacementMisconfigured` condition in the status of a placement?

Comment (haoqing0110) [reply pkg/plugins/tainttoleration/taint_toleration.go:47]: Yes, should consider this situation in PR https://github.com/open-cluster-management-io/placement/pull/56. Feel free to merge this first, I will try to cover it in https://github.com/open-cluster-management-io/placement/pull/56 .

File: pkg/plugins/tainttoleration/taint_toleration.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	// do validation on each toleration and return error if necessary
+	for _, toleration := range placement.Spec.Tolerations {
+		if len(toleration.Key) == 0 && toleration.Operator != clusterapiv1alpha1.TolerationOpExists {
+			return nil, errors.New("If the key is empty, operator must be Exists.\n")
+		}
+		if toleration.Operator == clusterapiv1alpha1.TolerationOpExists && len(toleration.Value) > 0 {
+			return nil, errors.New("If the operator is Exists, the value should be empty.\n")
+		}
+		if toleration.TolerationSeconds != nil && toleration.Effect != clusterapiv1.TaintEffectNoSelect && toleration.Effect != clusterapiv1.TaintEffectPreferNoSelect {
+			fmt.Println("Warning: TolerationSeconds would be ignored if Effect is not NoSelect/PreferNoSelect.")
```
Comment (qiujian16) [inline pkg/plugins/tainttoleration/taint_toleration.go:45]: klog here

File: pkg/plugins/tainttoleration/taint_toleration.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	// do validation on each toleration and return error if necessary
+	for _, toleration := range placement.Spec.Tolerations {
+		if len(toleration.Key) == 0 && toleration.Operator != clusterapiv1alpha1.TolerationOpExists {
+			return nil, errors.New("If the key is empty, operator must be Exists.\n")
+		}
+		if toleration.Operator == clusterapiv1alpha1.TolerationOpExists && len(toleration.Value) > 0 {
+			return nil, errors.New("If the operator is Exists, the value should be empty.\n")
+		}
+		if toleration.TolerationSeconds != nil && toleration.Effect != clusterapiv1.TaintEffectNoSelect && toleration.Effect != clusterapiv1.TaintEffectPreferNoSelect {
```
Comment (qiujian16) [inline pkg/plugins/tainttoleration/taint_toleration.go:44]: not sure why this if is needed, seems like a no-op

File: pkg/plugins/tainttoleration/taint_toleration.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		if isTolerated(taint, toleration) {
+			return true
+		}
+	}
+	return false
+}
+
+// isTolerated returns true if a taint is tolerated by the given toleration
+func isTolerated(taint clusterapiv1.Taint, toleration clusterapiv1alpha1.Toleration) bool {
+	if len(toleration.Effect) > 0 && toleration.Effect != taint.Effect {
```
Comment (qiujian16) [inline pkg/plugins/tainttoleration/taint_toleration.go:94]: do we need this `len(toleration.Effect) > 0`?

If so, it seems like it should have 

```go
if len(toleration.Effect) > 0 || len(toleration.Key) > 0 {
  return false
}
```

at the top to make the logic clearer

File: pkg/plugins/tainttoleration/taint_toleration.go (hunk 1 above)
Comment (qiujian16) [reply pkg/plugins/tainttoleration/taint_toleration.go:57]: +1

File: pkg/plugins/tainttoleration/taint_toleration.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
+package tainttoleration
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
+	clusterapiv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
```
Comment (haoqing0110) [inline pkg/plugins/tainttoleration/taint_toleration.go:10]: `placement` and `placementdecision` API updated to v1beta1, https://github.com/open-cluster-management-io/api/blob/main/cluster/v1beta1/types_placement.go, need to upgrade the version.

**Comment by elgnay** [comment]
Since the change has been committed by another PR #63, close this one. @AnthonyXu109, thanks for your contribution!

**Pull request closed without merging**